### PR TITLE
Fix mechanism to fetch filesystem overhead when the CDI resource has a different name

### PR DESCRIPTION
**Pull request by alromeros**
**What this PR does / why we need it**: 

The mechanism to obtain CDI's filesystem overhead in KubeVirt uses `cdi` as hardcoded key to check if the resource exists. This behavior conflicts when CDI has a custom name, such as `cdi-kubevirt-hyperconverged` in HCO.

This Pull Request aims to fix said mechanism, so the value is obtained independently of CDI's name.

**Which issue(s) this PR fixes:** 

https://bugzilla.redhat.com/show_bug.cgi?id=2097436

**Release note**:

```release-note
[Bug-fix]: Fix mechanism to fetch fs overhead when CDI resource has a different name
```

### CODE REVIEW

File: pkg/virt-controller/watch/vmi.go
Comment (ShellyKa13) [inline pkg/virt-controller/watch/vmi.go:2133]: I know I already commented that in a different PR but no need the else if you return in the first if :)

File: pkg/virt-controller/watch/vmi_test.go
Comment (ShellyKa13) [inline pkg/virt-controller/watch/vmi_test.go:2480]: I understand that it works because we dont check what the informer returns, just the it returned 1 but cdiInformer should return CDI instance and not CDIConfig, so you should create cdi1 and cdi2 of cdiv1.CDI{}

File: pkg/virt-controller/watch/vmi.go
Comment (ShellyKa13) [inline pkg/virt-controller/watch/vmi.go:2132]: I think its better to rephrase to "detected more then once CDI instance", also it should be %d and not %v
(sorry for the nitpicking)

File: pkg/virt-controller/watch/vmi_test.go
Comment (ShellyKa13) [inline pkg/virt-controller/watch/vmi_test.go:2508]: you can see here that you supposedly add a config but its really should have been an instance, so failing of not having cdi config although you supposedly created here 1 is not quit right

File: pkg/virt-controller/watch/vmi_test.go
Comment (ShellyKa13) [inline pkg/virt-controller/watch/vmi_test.go:2549]: here you should add an instance not the cfg you add to the cdiConfigInformer

**Review by ShellyKa13 — COMMENTED** [review]
some comments which needs adjustments but really great work at increasing the UT coverage of this function and not only adding the minimal needed for your fix

File: pkg/virt-controller/watch/vmi.go
Comment (awels) [inline pkg/virt-controller/watch/vmi.go:2133]: only nitpick is that we have a named error for `failedToFindCdi` and not one for multipleCdi. I would do either both named or neither named to be consistent.

Comment (alromeros) [reply pkg/virt-controller/watch/vmi.go:2133]: I agree but I see a minor problem there: `failedToFindCdi` is explicitly used in the error handling, so I think it makes sense that we use a named error just in that case. I can also create a new error for the other case, like `multipleCdiInstances`, but it wouldn't be used in the error-handling nor include the number of instances, which I like.

Comment (alromeros) [reply pkg/virt-controller/watch/vmi.go:2133]: Yeah, I agree. I'll change that :+1:

File: pkg/virt-controller/watch/vmi.go
Comment (rmohr) [inline pkg/virt-controller/watch/vmi.go:2094]: You can use `Reason(err)`. It can then look like this:

```
log.Log.V(3).Object(pvc).Reason(err).Infof("Failed to detect CDI, continuing with default filesystem overhead of 5.5%%")
```

**Review by rmohr — CHANGES_REQUESTED** [review]
One small nit.

**Comment by rmohr** [comment]
/lgtm
/approve

looks great, let's see if CI is happy too.

**Comment by rmohr** [comment]
/hold

until the test is inspected

**Comment by alromeros** [comment]
> Looks like one of the sig storage failures is real `Tests Suite: [sig-storage] [Serial]VirtualMachineClone Tests VM clone with more complicated VM with a simple clone` it is failing on all the storage lanes.

I think that test should be fixed once this PR is merged: https://github.com/kubevirt/kubevirt/pull/8059 (@ShellyKa13 correct me if I'm wrong). We also have a [Jira card](https://issues.redhat.com/browse/CNV-19647) to investigate why it's failing.

**Comment by ShellyKa13** [comment]
yes the PR #8059 should fix the failures which fails a lot of other PRs irrelevant to this one.. I think if this is the only thing blocking this PR from getting merged it should get a pass on that lanes

**Comment by rmohr** [comment]
> yes the PR #8059 should fix the failures which fails a lot of other PRs irrelevant to this one.. I think if this is the only thing blocking this PR from getting merged it should get a pass on that lanes

I did not follow the latest flaky-test developments closely. If it can be clearly identified as the same issue, feel free to remove the hold anytime.

**Pull request merged**
